### DISCUSSION TITLE
Use of FastAPI 100% without async

### DISCUSSION THREAD

**Discussion by MatthewScholefield** [fastapi/fastapi#3099] [Questions] [answered]
Consider a medium sized backend application built from the ground up using FastAPI. One of the first decisions that needs to be made is where to use async within your project. The intention in this discussion/question is to ask/figure out the following:

Is it reasonable to build your entire application using only sync libraries or in this case is Flask/greenlet a better choice?

**Motivation:**

`async` of course is infectious by design. This means that to maintain a unified codebase, either everything should be async or everything should be sync. Having a mix means a mixture of wrapping in `run_in_threadpool` and `asyncio.run` leading to extra overhead and the potential for bugs (calling sync code in async code). Another option is maintaining two sets of CRUD utility methods; one for sync and another for async, but this is of course much less than ideal since maintaining two of every function is a big hassle.

So, given that everything must be all async or sync, we have to compare them head to head. From what I've gathered, we have:

 - Async: Minimally developed libraries, and incredibly hard to debug errors
 - Sync: Unknown performance implication

There's also less room for race conditions with async code except that doesn't make much of a difference considering you might run multiple worker processes.

So, if my understanding of the pro and cons is correct, in any situation where you don't need to handle thousands of requests per second, sync would be better. Since I haven't seen this too often with FastAPI this conclusion felt wrong and I wanted to run it by some more experienced folks. What are your thoughts?

**Comment by MatthewScholefield** [comment, marked as answer]
Just an update on this, I've since converted an entire previously async codebase to regular sync code and everything seems to be running smoothly. The only caveat was that I had to use [asgiref](https://github.com/django/asgiref)'s `async_to_sync` and `sync_to_async` to handle websockets (for example `async_to_sync(websocket.accept)()`). Generally speaking, I've found asgiref to be immensely useful in preventing async from becoming too infectious.

**Edit** (update on the update): using `asgiref`'s async sync converter functions with websockets led to some strange errors. I ended up switching to creating a [`janus.Queue`](https://github.com/aio-libs/janus) per websocket request which I use to send messages between a sync function and the async websocket handler:

```python
@router.websocket('/wss/myhandler')
async def handle_my_websocket(websocket: WebSocket):
    await websocket.accept()
    queue = janus.Queue()
    send_json = queue.sync_q.put

    fut = asyncio.get_event_loop().run_in_executor(
        None, handle_my_websocket_sync, send_json
    )
    while True:
        val = await queue.async_q.get()
        if val is None:
            break
        await websocket.send_json(val)
    await fut

def handle_my_websocket_sync(send_json: Callable[[dict], None]):
    try:
        for _ in wait_for_application_update('some-param'):
            state = get_application_state('some-param')
            send_json(dict(state=state))
            if no_more_updates(state):
                return
    finally:
        send_update(None)

**Reply by aliriomambo** [reply]
Thank you for sharing your thoughts on this I had the same questions In mind.

**Comment by tiangolo** [comment]
I actually don't think that being async-first is the main benefit of FastAPI. Not even performance, but the speed of development, with data validation, automatic documentation, autocompletion and inline errors in the editor, etc.

Most of the apps I've seen with FastAPI are actually *not* async. But still, being able to go async for hot-paths can be useful, as it's all async to begin with.

About mixing async and sync code, I built Asyncer for that, to get better autocompletion, inline errors, etc. And there's a bunch of docs about doing these tricks. https://asyncer.tiangolo.com/

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
